README.md
# `flutter_im_list`

[ ![语言 中文](https://img.shields.io/badge/语言-中文-feb252.svg)](https://github.com/crazycodeboy/flutter_im_list/blob/master/README.cn.md)
[![Download](https://img.shields.io/badge/Download-ff69b4.svg) ](https://pub.dartlang.org/packages/flutter_im_list)
[ ![PRs Welcome](https://img.shields.io/badge/PRs-Welcome-brightgreen.svg)](https://github.com/crazycodeboy/flutter_im_list/pulls)
[ ![flutter_im_list release](https://img.shields.io/github/release/crazycodeboy/flutter_im_list.svg?maxAge=2592000?style=flat-square)](https://github.com/crazycodeboy/flutter_im_list/releases)
[![License MIT](http://img.shields.io/badge/license-MIT-orange.svg?style=flat)](https://raw.githubusercontent.com/crazycodeboy/flutter_im_list/master/LICENSE)

[`flutter_im_list`](https://github.com/crazycodeboy/flutter_im_list) is a high-performance, lightweight Flutter chat list package. It can help you quickly create a WeChat-like chat list effect.

## Content

- [Screenshots](#screenshots)
- [Examples](#examples)
- [Video tutorial](https://coding.imooc.com/class/672.html)
- [Getting started](#getting-started)
- [API](#api)
- [Contribution](#contribution)

## Screenshots

Overall | Long press | Inputting                                                                                                                   | GIF
---|---|-----------------------------------------------------------------------------------------------------------------------------|---
![flutter_im_list](https://raw.githubusercontent.com/crazycodeboy/flutter_im_list/master/example/Screenshots/1.pic.jpg)| ![flutter_im_list](https://raw.githubusercontent.com/crazycodeboy/flutter_im_list/master/example/Screenshots/2.pic.jpg)| ![flutter_im_list](https://raw.githubusercontent.com/crazycodeboy/flutter_im_list/master/example/Screenshots/3.pic.jpg) | ![flutter_im_list](https://raw.githubusercontent.com/crazycodeboy/flutter_im_list/master/example/Screenshots/4.gif)

## Examples

* [Examples](https://github.com/crazycodeboy/flutter_im_list/tree/master/example)

## [Video tutorial](https://coding.imooc.com/class/672.html)
Welcome to learn and communicate through [Video tutorial](https://coding.imooc.com/class/672.html).

## Getting started

### Step 1: Add dependencies

Run in the project root directory:

```bash
flutter pub add flutter_im_list
```
### Step 2: Initialize ChatController

```dart
@override
void initState() {
super.initState();
chatController = ChatController(
    initialMessageList: _messageList,
    timePellet: 60,
    scrollController: ScrollController());
}
```

### Step 3: Add ChatList to the layout

```dart
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: ChatList(
          chatController: chatController,
        ));
  }
```

### Step 4: Set initialization data

```dart
final List<MessageModel> _messageList = [
MessageModel(
    id: 1,
    content: "介绍下《ChatGPT + Flutter快速开发多端聊天机器人App》",
    ownerType: OwnerType.sender,
    createdAt: 1696142392000,
    avatar: 'https://o.devio.org/images/o_as/avatar/tx18.jpeg',
    ownerName: "Jack"),
MessageModel(
    id: 2,
    content:
        "当前ChatGPT应用如雨后春笋般应运而生，给移动端开发者也带来了极大的机会。本课程将整合ChatGPT与Flutter高级技术，手把手带你从0到1开发一款可运行在多端的聊天机器人App，帮助你抓住机遇，快速具备AI运用能力，成为移动端领域的AI高手。@https://coding.imooc.com/class/672.html",
    ownerType: OwnerType.receiver,
    createdAt: 1696142393000,
    avatar: 'https://o.devio.org/images/o_as/avatar/tx2.jpeg',
    ownerName: "ChatGPT"),
];
```
If not, you can assign _messageList to []。

>**To learn more see [Video tutorial](https://coding.imooc.com/class/672.html)**


## API

### IChatController
```dart
abstract class IChatController {
  /// Used to add message in message list.
  void addMessage(MessageModel message);

  /// Delete message.
  void deleteMessage(MessageModel message);
  /// Function for loading data while pagination.
  void loadMoreData(List<MessageModel> messageList);
}

```

### ChatController

```dart
class ChatController implements IChatController {
  /// Represents initial message list in chat which can be add by user.
  final List<MessageModel> initialMessageList;
  final ScrollController scrollController;

  ///Provide MessageWidgetBuilder to customize your bubble style.
  final MessageWidgetBuilder? messageWidgetBuilder;

  ///creation time group; unit second
  final int timePellet;
  List<int> pelletShow = [];

  ChatController({required this.initialMessageList,
    required this.scrollController,
    required this.timePellet,
    this.messageWidgetBuilder}) {
    for (var message in initialMessageList.reversed) {
      inflateMessage(message);
    }
  }
...
```

### ChatList

```dart
class ChatList extends StatefulWidget {
  /// Provides controller for accessing few function for running chat.
  final ChatController chatController;

  /// The amount of space by which to inset the children.
  final EdgeInsetsGeometry? padding;

  /// Called when the user taps this part of the material.
  final OnBubbleClick? onBubbleTap;

  /// Called when the user long-presses on this part of the material.
  final OnBubbleClick? onBubbleLongPress;
  /// support text select
  final HiSelectionArea? hiSelectionArea;

  const ChatList(
      {super.key,
      required this.chatController,
      this.padding,
      this.onBubbleTap,
      this.onBubbleLongPress,
      this.hiSelectionArea});

  @override
  State<ChatList> createState() => _ChatListState();
}
```
>**To learn more see [Video tutorial](https://coding.imooc.com/class/672.html)**

## Contribution

Issues are welcome. Please add a screenshot of you bug and a code snippet. Quickest way to solve
issue is to reproduce it in one of the examples.

Pull requests are welcome. If you want to change the API or do something big it is best to create an
issue and discuss it first.

---

**[MIT Licensed](https://github.com/crazycodeboy/flutter_im_list/blob/master/LICENSE)**


CHANGELOG.md
## 1.0.0

* v1.0.0


test/chat_message_test.dart
// import 'package:chat_message/chat_message.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:chat_message_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/core/chat_controller.dart
import 'dart:async';

import 'package:flutter/material.dart';

import '../models/message_model.dart';
import '../widget/message_widget.dart';

class ChatController implements IChatController {
  /// Represents initial message list in chat which can be add by user.
  final List<MessageModel> initialMessageList;
  final ScrollController scrollController;

  ///Provide MessageWidgetBuilder to customize your bubble style.
  final MessageWidgetBuilder? messageWidgetBuilder;

  ///creation time group; unit second
  final int timePellet;
  List<int> pelletShow = [];

  ChatController(
      {required this.initialMessageList,
      required this.scrollController,
      required this.timePellet,
      this.messageWidgetBuilder}) {
    for (var message in initialMessageList.reversed) {
      inflateMessage(message);
    }
  }

  /// Represents message stream of chat
  StreamController<List<MessageModel>> messageStreamController =
      StreamController();

  /// Used to dispose stream.
  void dispose() {
    messageStreamController.close();
    scrollController.dispose();
    initialMessageList.clear();
    pelletShow.clear();
  }

  ///ChatList is init ok
  void widgetReady() {
    if (!messageStreamController.isClosed) {
      messageStreamController.sink.add(initialMessageList);
    }
  }

  /// Used to add message in message list.
  @override
  void addMessage(MessageModel message) {
    //fix Bad state: Cannot add event after closing
    if (messageStreamController.isClosed) return;
    inflateMessage(message);
    // initialMessageList.add(message);
    //List反转后列是从底部向上展示，所以新来的消息需要插入到数据第0个位置
    initialMessageList.insert(0, message);
    messageStreamController.sink.add(initialMessageList);
    scrollToLastMessage();
  }

  @override
  void deleteMessage(MessageModel message) {
    if (messageStreamController.isClosed) return;
    initialMessageList.remove(message);
    pelletShow.clear();
    //时间的标记是从最久的消息开始标
    for (var message in initialMessageList.reversed) {
      inflateMessage(message);
    }
    messageStreamController.sink.add(initialMessageList);
  }

  /// Function for loading data while pagination.
  @override
  void loadMoreData(List<MessageModel> messageList) {
    //List反转后列是从底部向上展示，所以消息顺序也需要进行反转
    messageList = List.from(messageList.reversed);
    List<MessageModel> tempList = [...initialMessageList, ...messageList];
    //Clear record and redo
    pelletShow.clear();
    //时间的标记是从最久的消息开始标
    for (var message in tempList.reversed) {
      inflateMessage(message);
    }
    initialMessageList.clear();
    initialMessageList.addAll(tempList);
    if (messageStreamController.isClosed) return;
    messageStreamController.sink.add(initialMessageList);
  }

  /// Function to scroll to last messages in chat view
  void scrollToLastMessage() {
    //fix ScrollController not attached to any scroll views.
    if (!scrollController.hasClients) {
      return;
    }
    scrollController.animateTo(0,
        duration: const Duration(milliseconds: 200), curve: Curves.easeIn);
  }

  ///set showCreatedTime flag
  inflateMessage(MessageModel message) {
    int pellet = (message.createdAt / (timePellet * 1000)).truncate();
    if (!pelletShow.contains(pellet)) {
      pelletShow.add(pellet);
      message.showCreatedTime = true;
    } else {
      message.showCreatedTime = false;
    }
  }
}

abstract class IChatController {
  /// Used to add message in message list.
  void addMessage(MessageModel message);

  /// Delete message.
  void deleteMessage(MessageModel message);

  /// Function for loading data while pagination.
  void loadMoreData(List<MessageModel> messageList);
}


lib/models/message_model.dart
import 'package:flutter/widgets.dart';

enum OwnerType { receiver, sender }

OwnerType _of(String name) {
  if (name == OwnerType.receiver.toString()) {
    return OwnerType.receiver;
  } else {
    return OwnerType.sender;
  }
}

class MessageModel {
  /// Provides id
  final int? id;

  /// Avoid rebuilding the message widget when new incoming messages refresh the list.
  final GlobalKey key;

  /// Controls who is sending or receiving a message.
  /// Used to handle in which side of the screen the message
  /// will be displayed.
  final OwnerType ownerType;

  /// Name to be displayed with the initials.
  /// egg.: Higor Lapa will be H
  final String? ownerName;

  /// avatar url
  final String? avatar;

  /// The content to be displayed as a message.
  final String content;

  /// Provides message created time,milliseconds since.
  final int createdAt;

  ///Whether to display the creation time.
  bool showCreatedTime = false;

  MessageModel(
      {this.id,
      required this.ownerType,
      this.ownerName,
      this.avatar,
      required this.content,
      required this.createdAt})
      : key = GlobalKey();

  factory MessageModel.fromJson(Map<String, dynamic> json) => MessageModel(
        id: json["id"],
        content: json["content"],
        createdAt: json["createdAt"],
        ownerType: _of(json["ownerType"]),
        avatar: json["avatar"],
        ownerName: json["ownerName"],
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'content': content,
        'createdAt': createdAt,
        'ownerName': ownerName,
        //数据库存储不支持复合类型
        'ownerType': ownerType.toString(),
        'avatar': avatar
      };
}


lib/widget/message_widget.dart
import 'package:bubble/bubble.dart';
import 'package:flutter/material.dart';

import '../models/message_model.dart';
import '../util/wechat_date_format.dart';

typedef MessageWidgetBuilder = Widget Function(MessageModel message);
typedef OnBubbleClick = void Function(
    MessageModel message, BuildContext ancestor);

/// support text select
typedef HiSelectionArea = Widget Function(
    {required Text child, required MessageModel message});

class DefaultMessageWidget extends StatelessWidget {
  final MessageModel message;

  /// the font-family of the [content].
  final String? fontFamily;

  /// the font-size of the [content].
  final double fontSize;

  /// the size of the [avatar].
  final double avatarSize;

  /// the text-color of the [content].
  final Color? textColor;

  /// Background color of the message
  final Color? backgroundColor;
  final MessageWidgetBuilder? messageWidget;

  /// Called when the user taps this part of the material.
  final OnBubbleClick? onBubbleTap;

  /// Called when the user long-presses on this part of the material.
  final OnBubbleClick? onBubbleLongPress;

  final HiSelectionArea? hiSelectionArea;

  const DefaultMessageWidget(
      {required GlobalKey key,
      required this.message,
      this.fontFamily,
      this.fontSize = 16.0,
      this.textColor,
      this.backgroundColor,
      this.messageWidget,
      this.avatarSize = 40,
      this.onBubbleTap,
      this.onBubbleLongPress,
      this.hiSelectionArea})
      : super(key: key);

  double get contentMargin => avatarSize + 10;

  String get senderInitials {
    if (message.ownerName == null) return "";
    List<String> chars = message.ownerName!.split(" ");
    if (chars.length > 1) {
      return chars[0];
    } else {
      return message.ownerName![0];
    }
  }

  Widget get _buildCircleAvatar {
    var child = message.avatar is String
        ? ClipOval(
            child: Image.network(
              message.avatar!,
              height: avatarSize,
              width: avatarSize,
            ),
          )
        : CircleAvatar(
            radius: 20,
            child: Text(
              senderInitials,
              style: const TextStyle(fontSize: 16),
            ));
    return child;
  }

  @override
  Widget build(BuildContext context) {
    if (messageWidget != null) {
      return messageWidget!(message);
    }
    Widget content = message.ownerType == OwnerType.receiver
        ? _buildReceiver(context)
        : _buildSender(context);
    return Column(
      children: [
        if (message.showCreatedTime) _buildCreatedTime(),
        Padding(
          padding: const EdgeInsets.only(top: 10),
          child: content,
        ),
      ],
    );
  }

  Widget _buildReceiver(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.max,
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisAlignment: MainAxisAlignment.start,
      children: <Widget>[
        _buildCircleAvatar,
        Flexible(
          child: Bubble(
              margin: BubbleEdges.fromLTRB(10, 0, contentMargin, 0),
              stick: true,
              nip: BubbleNip.leftTop,
              color: backgroundColor ?? const Color.fromRGBO(233, 232, 252, 10),
              alignment: Alignment.topLeft,
              child: _buildContentText(TextAlign.left, context)),
        ),
      ],
    );
  }

  Widget _buildSender(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.max,
      mainAxisAlignment: MainAxisAlignment.end,
      crossAxisAlignment: CrossAxisAlignment.start,
      children: <Widget>[
        Flexible(
          child: Bubble(
              margin: BubbleEdges.fromLTRB(contentMargin, 0, 10, 0),
              stick: true,
              nip: BubbleNip.rightTop,
              color: backgroundColor ?? Colors.white,
              alignment: Alignment.topRight,
              child: _buildContentText(TextAlign.left, context)),
        ),
        _buildCircleAvatar
      ],
    );
  }

  Widget _buildContentText(TextAlign align, BuildContext context) {
    Widget text = Text(
      message.content,
      textAlign: align,
      style: TextStyle(
          fontSize: fontSize,
          color: textColor ?? Colors.black,
          fontFamily: fontFamily),
    );
    if (hiSelectionArea != null) {
      text = hiSelectionArea!.call(child: text as Text, message: message);
    }
    return InkWell(
        onTap: () =>
            onBubbleTap != null ? onBubbleTap!(message, context) : null,
        onLongPress: () => onBubbleLongPress != null
            ? onBubbleLongPress!(message, context)
            : null,
        child: text);
  }

  Widget _buildCreatedTime() {
    String showT = WechatDateFormat.format(message.createdAt, dayOnly: false);
    return Container(
      padding: const EdgeInsets.only(top: 10),
      child: Text(showT),
    );
  }
}


lib/util/wechat_date_format.dart
import 'package:intl/intl.dart';

class WechatDateFormat {
  /// millisecondsSinceEpoch:要格式化的日期；dayOnly:是只展示到天
  static String format(int millisecondsSinceEpoch, {bool dayOnly = true}) {
    //当前日期
    DateTime nowDate = DateTime.now();
    //传入的日期 millisecondsSinceEpoch
    DateTime targetDate =
        DateTime.fromMillisecondsSinceEpoch(millisecondsSinceEpoch);
    String prefix = "";
    if (nowDate.year != targetDate.year) {
      prefix = DateFormat("yyyy年M月d日").format(targetDate);
    } else if (nowDate.month != targetDate.month) {
      prefix = DateFormat("M月d日").format(targetDate);
    } else if (nowDate.day != targetDate.day) {
      if (nowDate.day - targetDate.day == 1) {
        prefix = '昨天';
      } else {
        prefix = DateFormat("M月d日").format(targetDate);
      }
    }
    if (prefix.isNotEmpty && dayOnly) {
      return prefix;
    }
    int targetHour = targetDate.hour;
    String returnTime = "", suffix = DateFormat("h:mm").format(targetDate);
    if (targetHour >= 0 && targetHour < 6) {
      returnTime = '凌晨';
    } else if (targetHour >= 6 && targetHour < 8) {
      returnTime = '早晨';
    } else if (targetHour >= 8 && targetHour < 11) {
      returnTime = '上午';
    } else if (targetHour >= 11 && targetHour < 13) {
      returnTime = '中午';
    } else if (targetHour >= 13 && targetHour < 18) {
      returnTime = '下午';
    } else if (targetHour >= 18 && targetHour <= 23) {
      returnTime = '晚上';
    }
    return '$prefix $returnTime$suffix';
  }

  static String formatYMd(int millisecondsSinceEpoch) {
    DateTime targetDate =
        DateTime.fromMillisecondsSinceEpoch(millisecondsSinceEpoch);
    return DateFormat("yyy/M/d").format(targetDate);
  }

  static String formatMd(int millisecondsSinceEpoch) {
    DateTime targetDate =
        DateTime.fromMillisecondsSinceEpoch(millisecondsSinceEpoch);
    return DateFormat("M-d").format(targetDate);
  }
}


lib/widget/chat_list_widget.dart
import 'package:flutter/material.dart';

import '../core/chat_controller.dart';
import '../models/message_model.dart';
import 'message_widget.dart';

class ChatList extends StatefulWidget {
  /// Provides controller for accessing few function for running chat.
  final ChatController chatController;

  /// The amount of space by which to inset the children.
  final EdgeInsetsGeometry? padding;

  /// Called when the user taps this part of the material.
  final OnBubbleClick? onBubbleTap;

  /// Called when the user long-presses on this part of the material.
  final OnBubbleClick? onBubbleLongPress;
  final HiSelectionArea? hiSelectionArea;

  const ChatList(
      {super.key,
      required this.chatController,
      this.padding,
      this.onBubbleTap,
      this.onBubbleLongPress,
      this.hiSelectionArea});

  @override
  State<ChatList> createState() => _ChatListState();
}

class _ChatListState extends State<ChatList> {
  ChatController get chatController => widget.chatController;

  MessageWidgetBuilder? get messageWidgetBuilder =>
      chatController.messageWidgetBuilder;

  ScrollController get scrollController => chatController.scrollController;

  Widget get _chatStreamBuilder {
    return StreamBuilder<List<MessageModel>>(
      stream: chatController.messageStreamController.stream,
      builder: (context, snapshot) {
        return snapshot.connectionState == ConnectionState.active
            ? ListView.builder(
                reverse: true,
                shrinkWrap: true,
                padding: widget.padding,
                controller: scrollController,
                itemCount: snapshot.data?.length ?? 0,
                itemBuilder: (BuildContext buildContext, int index) {
                  var model = snapshot.data![index];
                  return DefaultMessageWidget(
                    key: model.key,
                    message: model,
                    hiSelectionArea: widget.hiSelectionArea,
                    messageWidget: messageWidgetBuilder,
                    onBubbleTap: widget.onBubbleTap,
                    onBubbleLongPress: widget.onBubbleLongPress,
                  );
                },
              )
            : const Center(
                child: CircularProgressIndicator(),
              );
      },
    );
  }

  @override
  void initState() {
    super.initState();
    chatController.widgetReady();
  }

  @override
  Widget build(context) {
    //配合shrinkWrap: true使用，解决数据少的时候数据底部对齐的问题
    return Align(
      alignment: Alignment.topCenter,
      child: _chatStreamBuilder,
    );
  }

  @override
  void dispose() {
    chatController.dispose();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_im_list/core/chat_controller.dart';
import 'package:flutter_im_list/models/message_model.dart';
import 'package:flutter_im_list/widget/chat_list_widget.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'ChatMessage Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'ChatMessage Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, this.title = ""}) : super(key: key);
  final String title;

  @override
  MyHomePageState createState() => MyHomePageState();
}

class MyHomePageState extends State<MyHomePage> {
  int count = 0;
  final List<MessageModel> _messageList = [
    MessageModel(
        id: 1,
        content: "介绍下《ChatGPT + Flutter快速开发多端聊天机器人App》",
        ownerType: OwnerType.sender,
        createdAt: 1696142392000,
        avatar: 'https://o.devio.org/images/o_as/avatar/tx18.jpeg',
        ownerName: "Jack"),
    MessageModel(
        id: 2,
        content:
            "当前ChatGPT应用如雨后春笋般应运而生，给移动端开发者也带来了极大的机会。本课程将整合ChatGPT与Flutter高级技术，手把手带你从0到1开发一款可运行在多端的聊天机器人App，帮助你抓住机遇，快速具备AI运用能力，成为移动端领域的AI高手。@https://coding.imooc.com/class/672.html",
        ownerType: OwnerType.receiver,
        createdAt: 1696142393000,
        avatar: 'https://o.devio.org/images/o_as/avatar/tx2.jpeg',
        ownerName: "ChatGPT"),
  ];
  late ChatController chatController;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Column(
          children: [
            Expanded(
                child: ChatList(
              chatController: chatController,
            )),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: [
                ElevatedButton(
                    onPressed: _loadMore, child: const Text('loadMore')),
                ElevatedButton(onPressed: _send, child: const Text('Send'))
              ],
            ),
            const SizedBox(
              height: 20,
            )
          ],
        ));
  }

  @override
  void initState() {
    super.initState();
    chatController = ChatController(
        initialMessageList: _messageList,
        timePellet: 60,
        // messageWidgetBuilder: _diyMessageWidget,
        scrollController: ScrollController());
  }

  void _send() {
    chatController.addMessage(MessageModel(
        content: "I'm just a copy,${count++}",
        ownerType: OwnerType.sender,
        createdAt: DateTime.now().millisecondsSinceEpoch,
        avatar: 'https://o.devio.org/images/o_as/avatar/tx18.jpeg',
        ownerName: "Higor"));
    Future.delayed(const Duration(milliseconds: 2000), () {
      chatController.addMessage(
        MessageModel(
            content: "Nice",
            createdAt: DateTime.now().millisecondsSinceEpoch,
            ownerType: OwnerType.receiver,
            avatar: 'https://o.devio.org/images/o_as/avatar/tx2.jpeg',
            ownerName: "Bill Gates,${count++}"),
      );
    });
  }

  int _counter = 0;

  void _loadMore() {
    var tl = 1696142392000;
    tl = tl - ++_counter * 1000000;
    final List<MessageModel> messageList = [
      MessageModel(
          id: 3,
          content: "讲一个笑话",
          ownerType: OwnerType.sender,
          createdAt: tl,
          avatar: 'https://o.devio.org/images/o_as/avatar/tx18.jpeg',
          ownerName: "Jack"),
      MessageModel(
          id: 4,
          content: "那年愚人节，班上女神给了我一个奥利奥，我二话不说就咬了一口，才发现里面是牙膏。 ",
          ownerType: OwnerType.receiver,
          createdAt: tl,
          avatar: 'https://o.devio.org/images/o_as/avatar/tx2.jpeg',
          ownerName: "ChatGPT")
    ];
    chatController.loadMoreData(messageList);
  }

  ///Provide MessageWidgetBuilder to customize your bubble style.
  Widget _diyMessageWidget(MessageModel message) {
    return Container(
      margin: const EdgeInsets.only(top: 10, bottom: 10),
      padding: const EdgeInsets.only(top: 30, bottom: 30),
      decoration: BoxDecoration(
          color: message.ownerType == OwnerType.sender
              ? Colors.amberAccent
              : Colors.redAccent),
      child: Text('${message.ownerName}: ${message.content}'),
    );
  }
}
